### ISSUE TITLE
Password-protected backups and stored settings may have unencrypted secrets

### ISSUE THREAD

**Issue by Benno-K** [Authenticator-Extension/Authenticator#272]
**Platform:** Linux
 - Browser: Firefox
 - Browser Version: 60.4.0esr

When trying to export the backup of my 3 tokens one ends up unencrypted!
Looking into authenticator.json one token contains 
>    "encrypted": false,
and the secret is indeed not encrypted.

I experimented and added / deleted / re-added test-tokens and ended up with various results.

It seems to forget the password sometimes, but why then only some tokens get stored unencrypted is unclear to me. Anyway this is seems pretty unsafe to me. 

If I enter the password (and verification) once I would expect to need it for every single token and would expect nothing to be stored in cleartext nowhere from then on - until I remove the password.

**Comment by mymindstorm** [comment]
> If I enter the password (and verification) once I would expect to need it for every single token and would expect nothing to be stored in cleartext nowhere from then on - until I remove the password.

Password and unencrypted secrets are stored in memory after you put a password in. Did you set a password recently / how long ago did you set a password? If you set one again does it keep happening?

**Comment by mymindstorm** [comment]
Also if you run 
```javascript
chrome.storage.sync.get(d=>console.log(JSON.stringify(d,null,2)))
```
do any of them show up with encrypted: false? If you don't see any output replace `sync` with `local`

**Comment by Benno-K** [comment]
Oh, very quick response, thank you. I'm just going to bed so I will check and answer tomorrow - sorry.

**Comment by Benno-K** [comment]
So here are the steps, I just did them again.
Beforehand I deleted all existing tokens to have a defined testing environment.

1. Visited a page showing the QR-code of a token I use for testing
2. Entered a password in Security settings
3. Closed security settings
4. Opened them again to find the asterisks for the password still being shown (just to make sure)
5. Scanned the QR-code
6. Exported password protected settings
7. Found this
> {
>   "30b9df694cc62c6202211e99dddc7f01": {
>     "account": "TestToken2@freakmail.de",
>     "hash": "30b9df694cc62c6202211e99dddc7f01",
>     "issuer": "TestIssuer2",
>     "secret": "DEM2DEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMO",
>     "type": "totp",
>     "encrypted": false,
>     "index": 0,
>     "counter": 0
>   }
> }

On the other hand: If I then
8. Enter the password again - password used for testing is just 123
9. Exported password-protected settings
it get's encrypted
> {
>   "30b9df694cc62c6202211e99dddc7f01": {
>     "account": "TestToken2@freakmail.de",
>     "hash": "30b9df694cc62c6202211e99dddc7f01",
>     "issuer": "TestIssuer2",
>     "secret": "U2FsdGVkX1/npC2Rp4g7j6pFk9QaBvi4Kry6O0oWI8MPnjKpl+yEIJTgys3lTVA1o60zHElKvdxZcK05HXE0qRcLTw/ZTosT46EAXXRguSU=",
>     "type": "totp",
>     "encrypted": true,
>     "index": 0,
>     "counter": 0
>   }
> }

However, yesterday - in real life - I ended up with a mixture of entries, where "encrypted" was true for 3 of 4, but false for 1 of 4. So specifying the password just before saving seems not to cure all possible problems. And anyway if I specify a password once, it should not go away by itself under any circumstances (except I explicitly delete/unset it).

The token I'm using is
otpauth://totp/TestToken2%3ATestToken2%40freakmail.de?secret=DEM2DEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMO&issuer=TestIssuer2

I tried to paste it's QR-code into this writing, but it does not work, so at the end of this post I added the QR-code in base64 encoding.

I also notice that the password-asterisks are no longer shown as soon as I closed the Authenticator window once.

I tried to run the chrome-command
> chrome.storage.sync.get(d=>console.log(JSON.stringify(d,null,2)))

in the Firefox-debugger-console but all it gives me is

>ReferenceError: chrome is not defined[Weitere Informationen] debugger eval code:1:1 

so that was definitely the wrong place. Where do I run that?

Finally here's QR-code base64 encoded:
iVBORw0KGgoAAAANSUhEUgAAAJMAAACTAQMAAACwK7lWAAAABlBMVEUAAAD///+l2Z/dAAAAAnRS
TlP//8i138cAAAAJcEhZcwAACxIAAAsSAdLdfvwAAAIJSURBVEiJ3ZYxruQgEETbIiAbXwCJazjj
SuMLDPYF7CuRcQ0kLmAyAuTe8sz8+V/ahN6NdpETP0uou7ooTPz7ov+aHUQ+Njz3KXOgQcIKtwel
h1aHUyXiVcKC8VGdmtfKe6RZyOaA6lXhzH/AmDdSjEKkDH1M2LINwfKP3noY9JujeT/fmnYxrNMZ
X81Qzfg9yy52uHRziRxvGoUoESvB7pcGvMa8TcbLGPlqF013UiV8zbyPXSu0m7Z7SLDbKmLRjBGy
2VLzfk1ewM6r9XYnfDE3emnfy7AfEbRPQ7RrtSxhJ/EytTG0kdv49ksvK7HNNaOig9QOm4tYxYlM
Q6AbPcuRMUjOG8qpcE2bJey4XIZp59PZgyC/gJ0EpzSaLtMdjryEFQSJu/AYFdc0ShgzAgz75cVl
2JxFLPKijQ8IMPSRBgkrIQ3XgYbqsHnzIlYZb2tQG6W7fnm8l0E/D4YkqDDaR/s+hrRmhAEN2FK/
9OtmhOy0SHpIuNaXBr0Mc1u5zZH3qlDXLGHHlAghBKO560uRsIJxTTgZuKMQJ2YUMW6+5mNqc1AL
fe6ZLoaF/W44GRWARgl73jM4kfZ8nukiYc/wy4tWO/pgtYtYMA/K52RP14i+7pl+pi3jUq3mMfEq
ZD7CKbxed/Jb+16GPrSZK35/4NnPPLrYpR83QiTQJw962d/81/177BeUvoKMCyk8/gAAAABJRU5E
rkJggg==

**Comment by Benno-K** [comment]
OK, tried an additional thing.
1. Set password
2. Add additional QR-code

After that I had a look at
.mozilla/firefox-esr/\<censored\>.default/browser-extension-data/authenticator@mymindstorm/storage.js
where authenticator seems to store the data and unhappily found that the last token is not encrypted

>{"30b9df694cc62c6202211e99dddc7f01":{"account":"TestToken2@freakmail.de","hash":"30b9df694cc62c6202211e99dddc7f01","issuer":"TestIssuer2","secret":"U2FsdGVkX1/4/8g/IZUVxGFwPDu8tLxotCrTX/MWgYPZj9d4Z3WjSYFMC9rCEsCeKXG/ROkN6bQK0JAYJ7FQx+PF4Ajp5kwX8VN3jkccp60=","type":"totp","encrypted":true,"index":0,"counter":0},"34b57f718198dd300717154b8cfff250":{"account":"TestToken7@freakmail.de","hash":"34b57f718198dd300717154b8cfff250","issuer":"TestIssuer7","secret":"DEM7DEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMODEMO","type":"totp",**"encrypted":false**,"index":1,"counter":0}}

I guess this is the kind of information you tried me to get with "chrome.storage.sync.get(d=>console.log(JSON.stringify(d,null,2)))" - right?

**Comment by Benno-K** [comment]
The good news is:
If you once export a really all encrypted backup file (authenticator.json), then also all tokens are also encrypted in the store.


So the work-around to make sure everything is stored encrypted seems to be:

Whenever you add a token or change something else:
1. Set the password 
2. Immediately after that (do not allow the authenticator window to close in between) save a password-protected backup.
3. If no token is stored with '"encrypted": false' in it, then your secrets are stored safely and encrypted in the authenticator itself as well (and you could delete the backup).


But with the current version it is really likely to accidentally store unencrypted secrets, so I would love to see a fix, as I find the rest of the program just great (just discovered the neat qr-scan debugging page).

**Comment by Benno-K** [comment]
(this reply contained a link to some demo files which no longer exist)

**Comment by Benno-K** [comment]
(this reply contained a link to some demo video which no longer exists)

**Comment by mymindstorm** [comment]
Thanks for all the info, it helps a lot! I was able to reproduce, and will work on it when I have the time.

The data you found was what I was looking for. You can access the console via `about:debugging`

**Comment by Benno-K** [comment]
One more info:
When manually adding a token it is stored encrypted right from the beginning. Only when you scan a QR it gets stored unencrypted. So my guess is, that the scanning code somehow "forgets" to encrypt the secret.

**Comment by mymindstorm** [comment]
@Benno-K it tries to, but it was using an old variable for the password. Fix is already done, it will ship with 5.2.3

**Comment by Benno-K** [comment]
@mymindstorm: Thanks for quickly fixing it - I will remove the demo-page and the video now
